### DISCUSSION TITLE
Unable to solve environment when installing moose

### DISCUSSION THREAD

**Discussion by sSajjad90** [idaholab/moose#17094] [Q&A Getting Started] [answered]
Dear all,
I already had MOOSE on my system, and when I tried to update it, I got a chain of errors. So I decided to uninstall and re-install again. But I got stuck in the below trouble. It was in this state for more than 2 hours without a change. I interrupted it and try again, and the same problem.

Does anyone know what the problem is?

PS: My system memory is 32 Gb

(base) sajjad@sajjad-PC:~$ conda create --name moose moose-libmesh moose-tools
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment:\

**Comment by milljm** [comment, marked as answer]
I've seen lengthy solves among our continuous integration environment, but not on personal workstations. Can you try the following, and then attempt the install:

```bash
conda clean --all --yes
```

If the above fails, perhaps one module at a time:

```bash
conda create -n moose moose-libmesh
conda activate moose
conda install moose-tools
```

It is moose-tools which contains the most 'dependencies' Conda must solve for. And sometimes I see handing it one-package-at-a-time speeds things up. Let us know if none of these options work! I have more 'ideas'. Conda clean doesn't get everything.

**Reply by purepuremilk-zz** [reply]
I am having this similar issue, and I am installing moose on my personal laptop: solving the environment takes forever with "conda create --name moose moose-libmesh moose-tools". I tried to use the suggested approach by creating an empty environment first, activate it, and install moose-libmesh and moose-tools separately. I found the order of installing packages matters. Here are my commands: 

conda create -n moose
conda install -c idaholab moose-tools
conda install -c idaholab moose-libmesh

Here were what I found: If we installed moose-libmesh first and then moose-tool, solving enviroment for moose-tool would take a very long time since it was a large package. 

However, if we installed moose-tools first, which basically took no time, and then went to moose-libmesh, it still took a while to solve moose-libmesh, but in a much shorter time.

**Reply by milljm** [reply]
Conda solves can be a pain. moose-tools adds a lot of variety to the equation that Conda needs to solve for.

In the world of Conda, I try to be as non-explicit as possible to avoid potential shortcomings of our packages in the event another developer decides to pull a specific version from the Anaconda cloud. A version we might _explicitly_ state as a requirement. Thus, things work... most of the time. But that does come at a cost of solve time.

I am being very explicit with moose-libmesh. Because that package requires that we use the same compiler stack to build MOOSE (and allows you to build your Application). Hence, installing moose-libmesh is quick. Or rather, quicker than moose-tools.

**Reply by milljm** [reply]
@purepuremilk-zz thank you for this advice!

I have (will be) implementing this change into our Continuous Integration recipes on https://civet.inl.gov. To facilitate installing Conda packages separately. Specifically moose-tools first. As I too see a **_substantial speed increase_** doing so.

I will also modify our instructions to reflect this optimization: https://github.com/idaholab/moose/issues/17934

**Reply by milljm** [reply]
@purepuremilk-zz 

There also exists an interesting downgrading routine (explicit dependency somewhere?) for the following packages when installing moose-tools first, and then moose-libmesh (this being the only two packages requesting to be installed in a fresh environment on a Darwin machine):

```bash
The following packages will be REMOVED:

  glib-tools-2.68.2-he49afe7_0
  libgfortran5-9.3.0-h6c81a4c_22
  libglib-2.68.2-hd556434_0
  libnghttp2-1.43.0-h07e645a_0

The following packages will be SUPERSEDED by a higher-priority channel:

  glib                  conda-forge::glib-2.68.2-he49afe7_0 --> pkgs/main::glib-2.68.0-hdf23fa2_0

The following packages will be DOWNGRADED:

  curl                                    7.76.1-hb861fe1_2 --> 7.71.1-hbdca1c0_1
  dbus                                    1.13.6-ha13b53f_2 --> 1.13.6-h2f22bb5_0
  expat                                    2.4.1-he49afe7_0 --> 2.2.9-hb1e8313_2
  hdf4                                    4.2.15-hefd3b78_3 --> 4.2.13-h71d84a9_1004
  hdf5                           1.10.6-nompi_hc5d9132_1114 --> 1.10.6-nompi_h34ad4e8_1111
  krb5                                    1.19.1-hcfbf3a7_0 --> 1.17.2-h60d9502_0
  libblas                                  3.9.0-9_openblas --> 3.8.0-15_openblas
  libcblas                                 3.9.0-9_openblas --> 3.8.0-15_openblas
  libcurl                                 7.76.1-hf45b732_2 --> 7.71.1-he6690cf_1
  libcxx                                  11.1.0-habf9029_0 --> 11.0.0-h4c3b8ed_1
  libgfortran                       5.0.0-9_3_0_h6c81a4c_22 --> 4.0.0-7_5_0_h1a10cd1_22
  liblapack                                3.9.0-9_openblas --> 3.8.0-15_openblas
  libopenblas                      0.3.15-openmp_h5e1b9a4_1 --> 0.3.8-h3d69b6c_0
  libpng                                  1.6.37-h7cec526_2 --> 1.6.37-h2573ce8_0
  llvm-openmp                             11.1.0-hda6cdc1_1 --> 11.0.0-h73239a0_1
  matplotlib                           3.4.2-py37hf985489_0 --> 3.3.3-py37hf985489_0
  matplotlib-base                      3.4.2-py37hb018525_0 --> 3.3.3-py37hef07623_0
  openjpeg                                 2.4.0-h6e7aa92_1 --> 2.4.0-h6cbf5cd_0
  pandas                               1.2.4-py37hb23ed4d_0 --> 1.2.1-py37h010c265_0
  python                        3.7.10-h7728216_100_cpython --> 3.7.9-h6c3b2c9_0_cpython
  scipy                                1.6.3-py37h866764c_0 --> 1.5.3-py37h04d6967_0
  tbb                                     2020.2-h940c156_4 --> 2020.2-h926bf3e_3
```

So it stands to reason that perhaps moose-tools and moose-libmesh are at odds with one another. I'll continue investigating. But again, thank you for bringing this to my attention.

**Reply by milljm** [reply]
For reference. On Linux, the following seems to be inconsistent following a moose-tools installation:
```bash
The following packages will be SUPERSEDED by a higher-priority channel:

  libtiff               pkgs/main::libtiff-4.2.0-h85742a9_0 --> conda-forge::libtiff-4.1.0-hc3755c2_3
  zstd                     pkgs/main::zstd-1.4.5-h9ceee32_0 --> conda-forge::zstd-1.4.4-h3b9ef0a_2

The following packages will be DOWNGRADED:

  hdf5                           1.10.6-nompi_h6a2412b_1114 --> 1.10.5-nompi_h5b725eb_1114
  lcms2                                     2.12-hddcbb42_0 --> 2.11-hcbb858e_1
  libnetcdf                        4.7.4-nompi_h56d31a8_107 --> 4.6.2-h303dfb8_1003
  libpng                                  1.6.37-h21135ba_2 --> 1.6.37-hed695b0_0
  lz4-c                                    1.9.2-he1b5a44_3 --> 1.8.3-he1b5a44_1001
  openjpeg                                 2.4.0-hb52868f_1 --> 2.4.0-hf7af979_0
  pillow                               8.2.0-py37h4600e1f_1 --> 8.1.0-py37he6b4880_1
  vtk                                8.2.0-py37h2bd422c_218 --> 8.2.0-py37hfdee58b_203
```

**Comment by sSajjad90** [comment]
Many thanks for your response.  it works, but I got another problem. when I make the application (make -j4) There is the below error:

In file included from /home/sajjad/projects/moose/framework/build/unity_src/meshgenerators_Unity.C:25:0:
/home/sajjad/projects/moose/framework/src/meshgenerators/MeshGenerator.C: In member function 'std::unique_ptr<libMesh::MeshBase> MeshGenerator::generateInternal()':
/home/sajjad/projects/moose/framework/src/meshgenerators/MeshGenerator.C:92:62: error: no matching function for call to 'libMesh::MeshBase::get_info(int)'
     const auto mesh_info = mesh->get_info(/* verbosity = */ 2);
                                                              ^
In file included from /home/sajjad/projects/moose/framework/build/header_symlinks/MooseApp.h:26:0,
                 from /home/sajjad/projects/moose/framework/build/header_symlinks/MeshGenerator.h:14,
                 from /home/sajjad/projects/moose/framework/build/header_symlinks/SideSetsGeneratorBase.h:12,
                 from /home/sajjad/projects/moose/framework/build/header_symlinks/AllSideSetsByNormalsGenerator.h:12,
                 from /home/sajjad/projects/moose/framework/src/meshgenerators/AllSideSetsByNormalsGenerator.C:10,
                 from /home/sajjad/projects/moose/framework/build/unity_src/meshgenerators_Unity.C:2:
/home/sajjad/miniconda3/envs/moose/libmesh/include/libmesh/mesh_base.h:1194:15: note: candidate: std::__cxx11::string libMesh::MeshBase::get_info() const
   std::string get_info () const;
               ^~~~~~~~
/home/sajjad/miniconda3/envs/moose/libmesh/include/libmesh/mesh_base.h:1194:15: note:   candidate expects 0 arguments, 1 provided
Compiling C++ (in opt mode) /home/sajjad/projects/moose/framework/build/unity_src/executioners_Unity.C...
/home/sajjad/projects/moose/framework/build.mk:118: recipe for target '/home/sajjad/projects/moose/framework/build/unity_src/meshgenerators_Unity.x86_64-pc-linux-gnu.opt.lo' failed
make: *** [/home/sajjad/projects/moose/framework/build/unity_src/meshgenerators_Unity.x86_64-pc-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....




I should mention that I updated conda and moose but I receive this error. would you please let me know how could I fix it?

**Reply by milljm** [reply]
Performing `conda update`, or `conda install` commands should be followed by rebuilding your application after a cleaning. This is due to the library stack possibly getting updated (moose-libmesh was updated, as well as the compiler versions). There a few ways to 'clean' your application:

1. perform a `make clobber all` while in your application. Make sure to save (git commit) any changes you may have before doing so. Or:
2. perform a `git clean -xfd` while in the root of your application, and then again while inside the root of the `moose` directory. Again, making sure to save any changes you may have before doing so.

Once the above is complete, proceed to building your application.
